### PR TITLE
WS-760 combine course and program indices

### PR DESCRIPTION
**Pull request by rgraber**

### CODE REVIEW

**Comment by albemarle** [comment]
Amaze.

**Review by mikix — APPROVED** [review]

File: course_discovery/apps/course_metadata/algolia_proxy_models.py [hunk 1]
```diff
@@ -25,52 +27,159 @@ def get_owners(entry):
     return list(filter(lambda owner: owner['logoImageUrl'] is not None, all_owners))
 
 
-class AlgoliaProxyCourse(Course):
+def delegate_attributes(cls):
+    '''
+    Class decorator. For all Algolia fields, when my_instance.attribute is accessed, get the attribute off
+    my_instance.product rather than my_instance. This allows us to combine two different models into one index. If
+    my_instance.product doesn't have the attribute, attempts to access it will just return None.
```
Comment (albemarle) [inline course_discovery/apps/course_metadata/algolia_proxy_models.py:34]: Nice comment!

File: course_discovery/apps/course_metadata/algolia_proxy_models.py [hunk 2]
```diff
@@ -25,52 +27,159 @@ def get_owners(entry):
...
         if self.image:
             return getattr(self.image, 'url', None)
         return None
 
+    @property
     def owners(self):
         return get_owners(self)
 
+    @property
     def should_index(self):
```
Comment (albemarle) [inline course_discovery/apps/course_metadata/algolia_proxy_models.py:176]: This seems like a really important function that we should document in that wiki called "Debugging Algolia Problems" that doesn't exist yet.  ;-)

**Review by albemarle — APPROVED** [review]
Nice comments throughout.  Someone will appreciate them in the future.  :)

**Comment by jmbowman** [comment]
It looks like there should have been a migration to go along with this, work for the Django 2.2 upgrade is stumbling into that.  Can one of you generate that migration via `makemigrations` in its own PR so it doesn't trip developers up or perhaps cause deployment problems?

**Pull request merged**